src/app/functions/server/googleAnalytics.ts

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { getClientIdForSubscriber } from "../../../db/tables/google_analytics_clients";
+import { CONST_GA4_MEASUREMENT_ID } from "../../../constants";
+import { logger } from "./logging";
+
+/**
+ *   Send an event to GA4 backend using the GA Measurement Protocol.
+ *
+ * @param subscriberId - Monitor subscriber ID.
+ * @param eventName - the name of the ping (e.g. "purchase").
+ * @param eventParams - optional object containing key/value string pairs with additional information about this ping.
+ * @see https://developers.google.com/analytics/devguides/collection/protocol/ga4/sending-events?client_type=gtag
+ */
+export async function sendPingToGA(
+  subscriberId: number,
+  eventName: string,
+  eventParams: Record<string, string> = {},
+): Promise<void> {
+  const apiSecret = process.env.GA4_API_SECRET;
+  if (!apiSecret) {
+    throw new Error(
+      "No GA4 API secret is defined, cannot send backend events Google Analytics",
+    );
+  }
+
+  const gaClientInfo = await getClientIdForSubscriber(subscriberId);
+  if (!gaClientInfo) {
+    throw new Error(`No stored GA cookie for subscriber ${subscriberId}`);
+  }
+
+  const { client_id, cookie_timestamp } = gaClientInfo;
+
+  if (!client_id || !cookie_timestamp) {
+    throw new Error(
+      `No GA client_id found for subscriber ${subscriberId}, cannot send backend events to Google Analytics`,
```

Comment:
Small trick I learned from a former colleague: if you wrap the variable in start- and end-delimiters, it's easier to notice when a value is missing or malformed:

```suggestion
      `No GA client_id found for subscriber [${subscriberId}], cannot send backend events to Google Analytics`,
```

Reply:
Oh good idea - I didn't commit the suggestion because I found a few other places to fix this and pushed my branch.